### PR TITLE
FlatGeobuf: Upgrade to latest flatbuffers

### PR DESCRIPTION
**Pull request by bjornharrtell**

### CODE REVIEW

**Comment by bjornharrtell** [comment]
Looks like this has an effect on invalid data?

**Comment by rouault** [comment]
> Looks like this has an effect on invalid data?

just on invalid data ? Most flatgeobuf tests fail

**Comment by bjornharrtell** [comment]
> > Looks like this has an effect on invalid data?
> 
> just on invalid data ? Most flatgeobuf tests fail

Mmm that is concerning :( I will investigate.

**Comment by bjornharrtell** [comment]
I've been able to reproduce locally now. For reference these are the test results:

```
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_2 - AttributeError: 'NoneTy...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_3 - AttributeError: 'NoneTy...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_8 - AssertionError: assert ...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_9 - AssertionError: Verific...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_alldatatypes - AttributeErr...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_mixed - AttributeError: 'No...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_bool_short_float_binary - A...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_write_to_vsizip[spatial_index]
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_write_to_vsizip[no_spatial_index]
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_huge_number_of_columns - At...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_column_metadata - Attribute...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_editing - AttributeError: '...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_multi_geometries_with_empty[MULTIPOINT ((0 0), EMPTY)-MULTIPOINT ((0 0))]
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_multi_geometries_with_empty[MULTILINESTRING ((0 0,1 1), EMPTY)-MULTILINESTRING ((0 0,1 1))]
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_multi_geometries_with_empty[MULTIPOLYGON (((0 0,0 1,1 1,0 0)), EMPTY)-MULTIPOLYGON (((0 0,0 1,1 1,0 0)))]
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_multi_geometries_with_empty[GEOMETRYCOLLECTION (POINT (0 0), POINT EMPTY)-GEOMETRYCOLLECTION (POINT (0 0))]
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_ossfuzz_bug_29462 - Attribu...
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_read_invalid_geometries[data/flatgeobuf/invalid_polyhedralsurface_of_curvepolygon.fgb]
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_read_invalid_geometries[data/flatgeobuf/invalid_compoundcurve_non_contiguous_curves.fgb]
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_read_invalid_geometries[data/flatgeobuf/invalid_curvepolygon_linestring_three_points.fgb]
24: FAILED ogr/ogr_flatgeobuf.py::test_ogr_flatgeobuf_read_invalid_geometries[data/flatgeobuf/invalid_multisurface_of_polyhedralsurface.fgb]
24: =================== 21 failed, 10 passed, 1 warning in 1.41s ===================
```

**Comment by bjornharrtell** [comment]
I see many "Header failed consistency verification". This could possibly be alignment issues coming back to haunt.

**Comment by bjornharrtell** [comment]
Strange.. commenting out header validation makes all test pass, so feature buffer verification works but not header.

**Comment by bjornharrtell** [comment]
It does seem to have to do with alignment. The poly.fgb was written before https://github.com/OSGeo/gdal/commit/97738c0632e6780785114641d2bad9b7ec0482c6 but I've remade it with that fix in and I still cannot verify the header buffer with alignment check when the size prefix is not part of it. I've added the variants and specific test case upstream with https://github.com/flatgeobuf/flatgeobuf/commit/6225637f4685c4888065442f7ab1c313bef05dc1.

I'm not sure why this happens but the verifier must have become more picky. Luckily the feature data verifies fine including alignment check. I can't see why there would be a difference, at least after the mentioned patch that forces alignment on 8 bytes.

Long story short - I think we're no worse off by opting out of alignment check on header verification. I could make a copy of the buffer with the size prefix intact, which does seem to verify with alignment check but I don't see the benefit of doing so. So unless objections, if this becomes as green as it is for me locally, I would say this is ready to merge.

**Comment by rouault** [comment]
Tihs pull request causes a regression / security issue spotted by oss-fuzz, which doesn't occur in the release/3.5 branch where those changes aren't present.
With 
[clusterfuzz-testcase-minimized-fgb_fuzzer-4642513358553088.zip](https://github.com/OSGeo/gdal/files/8682877/clusterfuzz-testcase-minimized-fgb_fuzzer-4642513358553088.zip)
```
$ valgrind --leak-check=full ogrinfo clusterfuzz-testcase-minimized-fgb_fuzzer-4642513358553088 
==3118131== Memcheck, a memory error detector
==3118131== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==3118131== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==3118131== Command: ogrinfo clusterfuzz-testcase-minimized-fgb_fuzzer-4642513358553088
==3118131== 
==3118131== Invalid read of size 2
==3118131==    at 0x20D6B382: unsigned short flatbuffers::ReadScalar<unsigned short>(void const*) (base.h:408)
==3118131==    by 0x20D67468: flatbuffers::Table::GetOptionalFieldOffset(unsigned short) const (table.h:42)
==3118131==    by 0x20D674C7: flatbuffers::Table::VerifyOffset(flatbuffers::Verifier const&, unsigned short) const (table.h:135)
==3118131==    by 0x20D69BCC: FlatGeobuf::Header::Verify(flatbuffers::Verifier&) const (header_generated.h:551)
==3118131==    by 0x20D6E5F3: bool flatbuffers::Verifier::VerifyBufferFromStart<FlatGeobuf::Header>(char const*, unsigned long) (verifier.h:182)
==3118131==    by 0x20D6C179: bool flatbuffers::Verifier::VerifyBuffer<FlatGeobuf::Header>(char const*) (verifier.h:203)
==3118131==    by 0x20D6A36D: FlatGeobuf::VerifyHeaderBuffer(flatbuffers::Verifier&) (header_generated.h:700)
==3118131==    by 0x20D6326C: OGRFlatGeobufLayer::Open(char const*, _VSILFILE*, bool, bool) (ogrflatgeobuflayer.cpp:1551)
==3118131==    by 0x20D5889E: OGRFlatGeobufDataset::OpenFile(char const*, _VSILFILE*, bool) (ogrflatgeobufdataset.cpp:255)
==3118131==    by 0x20D58772: OGRFlatGeobufDataset::Open(GDALOpenInfo*) (ogrflatgeobufdataset.cpp:238)
==3118131==    by 0x5A4C889: GDALOpenEx (gdaldataset.cpp:3511)
==3118131==    by 0x11566E: main (ogrinfo.cpp:999)
==3118131==  Address 0x23bb758c is 44 bytes inside a block of size 45 alloc'd
==3118131==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==3118131==    by 0x4EB435D: VSIMalloc (cpl_vsisimple.cpp:561)
==3118131==    by 0x20D63163: OGRFlatGeobufLayer::Open(char const*, _VSILFILE*, bool, bool) (ogrflatgeobuflayer.cpp:1540)
==3118131==    by 0x20D5889E: OGRFlatGeobufDataset::OpenFile(char const*, _VSILFILE*, bool) (ogrflatgeobufdataset.cpp:255)
==3118131==    by 0x20D58772: OGRFlatGeobufDataset::Open(GDALOpenInfo*) (ogrflatgeobufdataset.cpp:238)
==3118131==    by 0x5A4C889: GDALOpenEx (gdaldataset.cpp:3511)
==3118131==    by 0x11566E: main (ogrinfo.cpp:999)
==3118131== 
ERROR 1: Header failed consistency verification
```

**Comment by bjornharrtell** [comment]
@rouault I interpret that report as a probable bug in the flatbuffers verifier itself. I will bring the fuzzed data upstream.. but meanwhile, do you think this is serious enough to warrant that I should revert this upgrade?
